GenericList/list.h

```diff
@@ -0,0 +1,240 @@
+        Resize(new_capacity);
+    }
+
+
+    void Print() const {
+        std::cout << "(";
+        for (size_t i = 0; i < count - 1; i++) {
+            std::cout << data[i] << ", ";
+        }
+        std::cout << data[count - 1] << ")\n";
+    }
+
+    //Take in any amount of arguments of any type, then unpack on element construction - allows for creating new data without checking for logic errors (the compiler and the element's constructor will take care of that)
+    template<typename... Args>
+    void Add(Args&&... args) {
+        if (capacity == count) Capacity(std::max(size_t(2) * capacity, size_t(1)));
+        new (data + count++) T(std::forward<Args>(args)...);
+    }
+
+    T* Find(const T& val) {
+        for (size_t i = 0; i < count; i++) {
+            if (data[i] == val) return data + i; //TODO: research why &data[i] might get overloaded
+        }
+        return nullptr;
+    }
+    const T* Find(const T& val) const {
+        for (size_t i = 0; i < count; i++) {
+            if (data[i] == val) return data + i; //TODO: research why &data[i] might get overloaded
+        }
+        return nullptr;
+    }
+
+    template <typename Predicate>
+    T* FindIf(Predicate&& pred) {
+        for (auto ptr = begin(); ptr < end(); ptr++) {
+            if (pred(*ptr)) return ptr; //TODO: research why &data[i] might get overloaded
+        }
+        return nullptr;
+    }
+
+    template <typename Predicate>
+    const T* FindIf(Predicate&& pred) const {
+        for (auto ptr = begin(); ptr < end(); ptr++) {
+            if (pred(*ptr)) return ptr; //TODO: research why &data[i] might get overloaded
+        }
+        return nullptr;
+    }
+
+    void RemoveAt(size_t index) {
+        if (index < 0 || index >= count) throw std::out_of_range(string("Cannot remove element at out_of_range index: ") + to_string(index) + string("."));
+
+        std::move(data + index + 1, data + count, data + index);
+        data[--count].~T();
+    }
+
+
+    //Returns true if a relevant element exists, and removes it.
+    size_t Remove(const T& val) {
+        //Lazy, simpler way - when changing behaviour, just change it in RemoveIf:
+        return RemoveIf([&](const auto& e) { return e == val; } );
+    }
+    
+    template <typename Predicate>
+    size_t RemoveIf(Predicate&& pred) {
+        //idea: double pointers, picker points to next element to check,
+        //                       placer points to next available space
+
+        auto placer = FindIf(pred);
+        if (placer == nullptr) return 0;
+
+        for (auto picker = placer + 1; picker < end(); ++picker) {
+            if (!pred(*picker)) {
+                *placer = std::move(*picker); //*picker now contains irrelevant instantiated data
+                ++placer;
+            }
+        }
+
+        for (auto k = placer; k < end(); ++k) {
+            k->~T(); //destructs meaningless data from matched indexes moved to end
+        }
```

Comment:
I don't think this will make a difference at runtime tbh, but still useful to know. Look up `if constexpr`, 
```suggestion
        if constexpr (!std::is_trivially_destructible_v<T>)
        {
            for (auto k = placer; k < end(); ++k) {
                k->~T(); //destructs meaningless data from matched indexes moved to end
            }
        }
```